crates/flightrepl/src/completer.rs

```diff
@@ -0,0 +1,502 @@
+use crate::EditorHelper;
+use arrow_flight::flight_service_client::FlightServiceClient;
+use datafusion::arrow::array::{Array, StringArray};
+use rustyline::Context;
+use rustyline::completion::{Completer, Pair};
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::{RwLock, oneshot};
+use tokio::time::interval;
+use tonic::transport::Channel;
+
+#[rustfmt::skip]
+const SQL_KEYWORDS: &[&str] = &[
```

Comment:
Datafusion has a [list of keywords](https://docs.rs/datafusion/latest/datafusion/logical_expr/sqlparser/dialect/keywords/constant.ALL_KEYWORDS.html) from it's SQL parser.

Reply:
Along these lines, I think it would be easier to use the [sqlparser](https://docs.rs/sqlparser/latest/sqlparser/) crate to incrementally parse the query as the user enters it. You would be able to transform an assertion like this: 
https://github.com/spiceai/spiceai/pull/6589/files#diff-aba255444129f489045ef42af55102c0d3166a1edf4a629c0cc38cb73f469992R141-R143

To one that folds over the AST "so-far" and realizes "oh, there's a select, but no table ident!".

Reply:
I tried to play with this idea, but the problem is it only works when we have a completely valid query. 

For example it fails to process these strings, where currentl implementation properly handles them
```
SELE
SELECT
SELECT * FROM table1 inner 
SELECT * FROM table1 inner jo
SELECT * FROM table1 inner join 
```

So even example ` there's a select, but no table ident` won't work.

Given this limitation I feel like it won't add much benefit. I think the current implementation is a good balance between user experience (handles most cases) and complexity (not too complex).

Reply:
I agree, if we can't get the benefits of `sqlparser`, using `datafusion::sql::sqlparser::keywords::ALL_KEYWORDS` may be the most we can do.